doc/decisions/array.md

```diff
@@ -40,11 +38,30 @@ myarray/#5 = value5
   array = #5
 ```
 
+To lookup an array, first do `ksLookup(ks,"/myarray")` on the parent.
+With the last index you get from its metadata `array`, iterate over the children.
+A cascading lookup on every individual child is also needed to make sure that overrides on individual
+elements are respected.
```

Comment:
A helper function for this might be useful. This kind of lookup is sadly also quite expensive.

Reply:
> A helper function for this might be useful.

Yes, definitely. Or bindings you probably even want that you can access it as an array of the programming language.

> This kind of lookup is sadly also quite expensive.

I never saw any application where more than 3% was spent on lookup (even with several fallbacks/overrides).

Did you implement this kind of lookup already in the high-level API?

Reply:
> I never saw any application where more than 10% was spent on lookup (even with several fallbacks/overrides).

10% of what?

It's also not a good idea to measure this in percentages. Of course optimizations should be performed on the bottlenecks (right now probably some part of `kdbGet`/`kdbSet`). But if we have another reason to change arrays, improving performance a long the way would still be nice.

BTW the additional lookup also has additional cost in terms of developer time and code complexity. That could also be called expensive.

> Did you implement this kind of lookup already in the high-level API?

The high-level API does some handling of `array` metadata. It will have to be seen, if that turns out to be correct or not, once this decision is decided.

Reply:
>  10% of what?

E.g. CPU time. But actually of nearly any metric I can think of, as lookup e.g. has only a few simple loops and does only one allocation :wink:

> It's also not a good idea to measure this in percentages.

It is even a worse idea to speculate without any measurement.

> BTW the additional lookup also has additional cost in terms of developer time and code complexity. That could also be called expensive.

The sword has two edges: that every key the application wants to use needs a lookup is a totally simple rule, easy to follow and provides consistent behavior.

Reply:
I added this proposal now here 16fe87f7687a0f82dae2b757cf405f1d8b03bc33

Reply:
> It is even a worse idea to speculate without any measurement.

There is no need for measurement to confirm that fewer lookups is faster. It might not be noticeable, but like I said performance is not my primary goal but just a side effect.

> has only a few simple loops and does only one allocation 

`ksLookup` can be recursive and use more than one allocation. It all depends on what lookup you are doing.

> every key the application wants to use needs a lookup is a totally simple rule

For applications this is fine, but I'm more concerned about plugins and libraries that need to deal with arrays. They also need to do lookups when a simple iteration over a `KeySet` may have been enough. In such a case the code is suddenly a lot more complicated.

Reply:
> ksLookup can be recursive and use more than one allocation.

Which allocation do you mean? I mean the one allocation of the lookup key (which can even be avoided by ksLookupByName). So recursive calls do not change the number of this allocation, as the allocation happened outside.

> They also need to do lookups when a simple iteration over a KeySet may have been enough.

When they are not concerned about other namespaces, the simple iteration should be enough anyway.

Reply:
> Which allocation do you mean?

There are many `keySetNamespace` calls in the helper functions (e.g. `elektraLookupByCascading`). All of these result in allocations. Total memory use will be pretty constant, but there could still be a lot of `realloc`.

There is also `elektraLookupBySpecLinks` which has a `keyNew`.